core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -331,17 +418,22 @@ class LogManager(val logDirs: Array[File],
           log.truncateTo(truncateOffset)
           if (needToStopCleaner)
             cleaner.maybeTruncateCheckpoint(log.dir.getParentFile, topicPartition, log.activeSegment.baseOffset)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to truncate log for $topicPartition due to IOException", e)
+            handleLogDirFailure(log.dir.getParent)
```

Comment:
I have updated the patch so that the partitions will be removed from the replica fetcher if truncation fails.